### PR TITLE
Initial non-optimized support for automatically chunking responses

### PR DESCRIPTION
**Pull request by halter73**
#40

### CODE REVIEW

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 1]
```diff
@@ -235,6 +236,12 @@ private void FireOnCompleted()
             try
             {
                 await Application.Invoke(this).ConfigureAwait(false);
+
+                if (_autoChunk)
+                {
+                    WriteChunkPrefix(numOctets: 0);
```
Comment (lodejard) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: The prefix and suffix combo can probably just be a constant.

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 2]
```diff
@@ -247,11 +254,53 @@ private void FireOnCompleted()
             }
         }
 
-
         public void Write(ArraySegment<byte> data, Action<Exception, object> callback, object state)
         {
             ProduceStart(immediate: false);
-            SocketOutput.Write(data, callback, state);
+
+            if (_autoChunk)
```
Comment (lodejard) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: Watch out for zero-length writes - when chunking is active they can look like an end-of-stream in the middle of the response.

**Comment by lodejard** [comment]
:shipit: - with two notes. only the one about zero-length writes is critical

**Comment by davidfowl** [comment]
Where are the tests?

**Comment by lodejard** [comment]
oh snap

**Comment by halter73** [comment]
@davidfowl I'll be enabling [this ServerTest](https://github.com/aspnet/ServerTests/blob/e48cf0a91bd9915261301565cd8d99ab5b0f5961/test/ServerComparison.FunctionalTests/ResponseTests.cs#L65) once this is merged. I'll add an EngineTest too. Including one with a zero length write.

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 3]
```diff
@@ -265,11 +271,74 @@ private void FireOnCompleted()
...
+            {
+                if (data.Count == 0)
+                {
+                    callback(null, state);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: named parameters..........

File: src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs [hunk 4]
```diff
@@ -265,11 +271,74 @@ private void FireOnCompleted()
...
+        private void WriteChunkPrefix(int numOctets)
+        {
+            var numOctetBytes = CreateAsciiByteArraySegment(numOctets.ToString("x") + "\r\n");
+
+            SocketOutput.Write(numOctetBytes,
+                    (error, _) =>
+                    {
+                        if (error != null)
+                        {
+                            Trace.WriteLine("WriteChunkPrefix" + error.ToString());
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs]: :disappointed:

**Comment by davidfowl** [comment]
:shipit:

**Pull request merged**